lib/Dialect/Moore/MooreOps.cpp

```diff
@@ -1417,6 +1417,110 @@ void ClassNewOp::getEffects(
   effects.emplace_back(MemoryEffects::Allocate::get());
 }
 
+LogicalResult ClassUpcastOp::verify() {
+  // 1) Type checks.
+  auto srcTy = dyn_cast<ClassHandleType>(getOperand().getType());
+  if (!srcTy)
+    return emitOpError() << "operand must be !moore.class<...>; got "
+                         << getOperand().getType();
+
+  auto dstTy = dyn_cast<ClassHandleType>(getResult().getType());
+  if (!dstTy)
+    return emitOpError() << "result must be !moore.class<...>; got "
+                         << getResult().getType();
+
+  if (srcTy == dstTy)
+    return success();
+
+  auto *op = getOperation();
+
+  auto *srcDeclOp =
+      mlir::SymbolTable::lookupNearestSymbolFrom(op, srcTy.getClassSym());
+  auto *dstDeclOp =
+      mlir::SymbolTable::lookupNearestSymbolFrom(op, dstTy.getClassSym());
+  if (!srcDeclOp || !dstDeclOp)
+    return emitOpError() << "failed to resolve class symbol(s): src="
+                         << srcTy.getClassSym()
+                         << ", dst=" << dstTy.getClassSym();
+
+  auto srcDecl = dyn_cast<ClassDeclOp>(srcDeclOp);
+  auto dstDecl = dyn_cast<ClassDeclOp>(dstDeclOp);
+  if (!srcDecl || !dstDecl)
+    return emitOpError()
+           << "symbol(s) do not name `moore.class.classdecl` ops: src="
+           << srcTy.getClassSym() << ", dst=" << dstTy.getClassSym();
+
+  auto cur = srcDecl;
+  while (cur) {
+    if (cur == dstDecl)
+      return success(); // legal upcast: dst is src or an ancestor
+
+    auto baseSym = cur.getBaseAttr();
+    if (!baseSym)
+      break;
+
+    auto *baseOp = mlir::SymbolTable::lookupNearestSymbolFrom(op, baseSym);
+    cur = llvm::dyn_cast_or_null<ClassDeclOp>(baseOp);
+  }
+
+  return emitOpError() << "cannot upcast from " << srcTy.getClassSym() << " to "
+                       << dstTy.getClassSym()
+                       << " (destination is not a base class)";
+}
+
+LogicalResult ClassPropertyRefOp::verify() {
+  // The operand is constrained to ClassHandleType in ODS; unwrap it.
+  Type instTy = getInstance().getType();
+  auto handleTy = dyn_cast<moore::ClassHandleType>(instTy);
+  if (!handleTy)
+    return emitOpError() << "instance must be a !moore.class<@C> value, got "
+                         << instTy;
+
+  // Extract the referenced class symbol from the handle type.
+  SymbolRefAttr classSym = handleTy.getClassSym();
+  if (!classSym)
+    return emitOpError("instance type is missing a class symbol");
+
+  // Resolve the class symbol starting from the nearest symbol table.
+  Operation *clsSym =
+      mlir::SymbolTable::lookupNearestSymbolFrom(getOperation(), classSym);
+  if (!clsSym)
+    return emitOpError("referenced class symbol `")
+           << classSym << "` was not found";
```

Comment:
The `ClassPropertyRefOp` probably also needs to be a symbol user such that you can verify that the referenced field actually exists in the class.